In [1]:
import pandas as pd
import numpy as np

In [26]:
data = pd.read_csv('Datasets/train.csv',nrows=200)

In [27]:
data.rename(columns={'question':'prompt','sql':'completion'}, inplace=True)

In [4]:
 pip install openai

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [28]:
data['prompt'] = data['prompt'].str.strip()
data['completion'] = data['completion'].str.strip()

In [29]:
data['prompt'] = data['prompt'] + "\n\n###\n\n"
#
data['completion'] = " " + data['completion'] + " STOP"

In [30]:
data.to_json("p_and_c.jsonl",orient='records',lines=True)

In [31]:
import os
os.environ['OPENAI_API_KEY'] = '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'

In [32]:
openai.api_key = "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"

In [33]:
!openai tools fine_tunes.prepare_data -f p_and_c.jsonl

Analyzing...

- Your file contains 200 prompt-completion pairs
- There are 1 duplicated prompt-completion sets. These are rows: [161]
- More than a third of your `completion` column/key is uppercase. Uppercase completions tends to perform worse than a mixture of case encountered in normal language. We recommend to lower case the data if that makes sense in your domain. See https://beta.openai.com/docs/guides/fine-tuning/preparing-your-dataset for more details
- All prompts end with suffix `\n\n###\n\n`
- All completions start with prefix ` SELECT `. Most of the time you should only add the output data into the completion, without any prefix
- All completions end with suffix ` STOP`

Based on the analysis we will perform the following actions:
- [Recommended] Remove 1 duplicate rows [Y/n]: ^C



Analyzing...

- Your file contains 56355 prompt-completion pairs
- There are 189 duplicated prompt-completion sets. These are rows: [161, 1238, 1356, 2122, 2283, 2338, 2505, 2847, 3020, 3114, 3134, 3215, 3217, 3463, 4145, 5212, 5213, 6124, 6170, 6555, 6739, 7328, 7448, 7864, 7944, 8151, 8572, 8582, 8845, 9099, 9660, 9912, 10038, 10255, 10727, 10795, 11050, 11273, 11408, 11824, 11920, 12091, 12137, 12270, 14124, 14302, 14377, 14604, 15031, 15094, 15632, 16265, 16374, 17128, 17152, 17199, 17596, 17955, 18025, 18195, 19014, 19488, 19496, 19756, 20242, 20986, 21148, 22346, 22351, 22854, 23478, 23517, 24731, 24935, 25195, 25429, 25549, 25883, 25937, 26994, 27131, 27132, 27272, 27706, 27832, 28693, 29652, 29923, 30017, 30101, 30535, 30827, 31096, 31475, 32150, 32275, 32846, 32946, 33273, 33726, 34621, 35403, 35455, 35616, 35617, 36280, 36325, 37283, 37325, 37371, 37394, 37489, 37961, 38226, 38400, 38405, 38685, 38873, 39419, 39637, 40017, 40088, 40089, 40298, 40368, 40420, 40481, 40968, 41115, 41124, 41215, 41218, 41277, 41551, 41599, 41741, 42108, 42111, 42396, 42402, 42643, 42823, 43121, 43166, 43378, 43384, 43443, 43740, 43854, 44416, 44760, 44884, 45376, 45842, 46147, 46375, 46378, 46633, 46676, 46714, 47147, 47523, 47623, 48056, 48205, 48732, 49314, 49415, 49978, 50405, 51494, 51656, 51658, 51983, 53475, 53592, 53602, 54065, 54741, 54801, 54837, 54865, 55156, 55212, 55320, 55494, 55529, 55762, 56156]
- More than a third of your `completion` column/key is uppercase. Uppercase completions tends to perform worse than a mixture of case encountered in normal language. We recommend to lower case the data if that makes sense in your domain. See https://beta.openai.com/docs/guides/fine-tuning/preparing-your-dataset for more details
- All prompts end with suffix `\n\n###\n\n`
- All completions start with prefix ` SELECT `. Most of the time you should only add the output data into the completion, without any prefix
- All completions end with suffix ` STOP`

Based on the analysis we will perform the following actions:
- [Recommended] Remove 189 duplicate rows [Y/n]: Y
- [Recommended] Lowercase all your data in column/key `completion` [Y/n]: Y
/home/rejeesh/.local/lib/python3.8/site-packages/openai/validators.py:448: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x[column] = x[column].str.lower()
- [Recommended] Remove prefix ` SELECT ` from all completions [Y/n]: n


Your data will be written to a new JSONL file. Proceed [Y/n]: y

Wrote modified file to `p_and_c_prepared.jsonl`
Feel free to take a look!

Now use that file when fine-tuning:
> openai api fine_tunes.create -t "p_and_c_prepared.jsonl"

After you’ve fine-tuned a model, remember that your prompt has to end with the indicator string `\n\n###\n\n` for the model to start generating completions, rather than continuing with the prompt. Make sure to include `stop=[" stop"]` so that the generated texts ends at the expected place.
Once your model starts training, it'll approximately take 2.06 days to train a `curie` model, and less for `ada` and `babbage`. Queue will approximately take half an hour per job ahead of you.

In [35]:
!openai api fine_tunes.create -t "p_and_c_prepared.jsonl" -m 'davinci'

Upload progress: 100%|████████████████████| 33.3k/33.3k [00:00<00:00, 11.4Mit/s]
Uploaded file from p_and_c_prepared.jsonl: file-Gb4CNfBEehrRfNIXm4gg7AWA
Created fine-tune: ft-QkK8J86U4IHUTqGWf6UcNbAe
Streaming events until fine-tuning is complete...

(Ctrl-C will interrupt the stream, but not cancel the fine-tune)
[2023-03-02 23:51:19] Created fine-tune: ft-QkK8J86U4IHUTqGWf6UcNbAe

Stream interrupted (client disconnected).
To resume the stream, run:

  openai api fine_tunes.follow -i ft-QkK8J86U4IHUTqGWf6UcNbAe



In [43]:
!openai api fine_tunes.follow -i ft-QkK8J86U4IHUTqGWf6UcNbAe

[2023-03-02 23:51:19] Created fine-tune: ft-QkK8J86U4IHUTqGWf6UcNbAe
[2023-03-02 23:57:03] Fine-tune costs $0.79
[2023-03-02 23:57:03] Fine-tune enqueued. Queue number: 6
[2023-03-02 23:57:12] Fine-tune is in the queue. Queue number: 5
[2023-03-02 23:57:13] Fine-tune is in the queue. Queue number: 4
[2023-03-02 23:59:44] Fine-tune is in the queue. Queue number: 3
[2023-03-02 23:59:53] Fine-tune is in the queue. Queue number: 2
[2023-03-03 00:01:02] Fine-tune is in the queue. Queue number: 1
[2023-03-03 00:02:20] Fine-tune is in the queue. Queue number: 0
[2023-03-03 00:05:04] Fine-tune started
[2023-03-03 00:07:32] Completed epoch 1/4
[2023-03-03 00:08:30] Completed epoch 2/4
[2023-03-03 00:09:28] Completed epoch 3/4
[2023-03-03 00:10:26] Completed epoch 4/4
[2023-03-03 00:11:01] Uploaded model: davinci:ft-dqlabs-ai-2023-03-02-18-41-01
[2023-03-03 00:11:03] Uploaded result file: file-cqsrTDD2hdBaltZ2tCCItwMs
[2023-03-03 00:11:03] Fine-tune succeeded

Job complete! Status: succeeded 🎉
T

In [12]:
prompt = "Retrieve the name and total amount of the customer who placed the order with the highest total amount\n\n###\n\n"

In [47]:
 openai.api_key = "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"

 response = openai.Completion.create(
   model="davinci:ft-dqlabs-ai-2023-03-02-18-41-01",
   prompt=prompt,
   temperature=0.7,
   max_tokens=256,
   top_p=1,
   frequency_penalty=0,
   presence_penalty=0,
   stop= [" STOP"]
 )


In [48]:
print(response['choices'][0]['text'])

 select name from table where total amount = highest total amount stop stop


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###

 customer name = stop stop


###

 stop stop ###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###


###



